text/0016-internationalization.md

```diff
@@ -0,0 +1,117 @@
+    
+5.  Possibly find ways to reuse OpenMRS module translations found in the `message.properties` file.
+
+7.  Update mechanism: Once translations are marked as complete, create a mechanism to publish the translated strings to a github repo. 
+    
+
+## Reason for decision
+
+- Most distributions are deployed to areas with no to intermittent connectivity, thus the translations module should be completely offline.
+
+- English is the lingua franca and spoken in most parts of the world.
+
+- Reasons for choosing Transifex:
+	 - Transifex is the most popular internationalization SaaS and offers tools that would make the translations workflow smoother.
+	 - The community is already familiar using transifex for translation services.
+
+
+## Alternatives
+- No translation service at all and default to English for all frontend modules.
+-  Use another service provider built specifically for frontend applications such as [react-i18next](https://github.com/i18next/react-i18next)
+- We could store the translations with each `esm` module instead of having a translations `esm` module.
+- Repo to handle translations service: we could have the translations handled by `openmrs-esm-api`
+
+
+## Common practices (not enforced)
+
+
+### Definitions
+
+Messages - OpenMRS historically used this word to refer to the content that would be replaced.
+
+  
+### Structure of Message File
+
+Here’s an example of how message.properties file looks like:
+
+[https://github.com/openmrs/openmrs-module-legacyui/blob/master/api/src/main/resources/messages.properties](https://github.com/openmrs/openmrs-module-legacyui/blob/master/api/src/main/resources/messages.properties)
+
+
+### Existing API
+
+Here is a [link](https://github.com/openmrs/openmrs-core/blob/master/api/src/main/java/org/openmrs/messagesource/MessageSourceService.java) to the github repo for the interface to the java MessageSourceService in OpenMRS core. It has four methods:
+
+    public String getMessage(String s);
+    
+    public MutableMessageSource getActiveMessageSource();
+    
+    public  void  setActiveMessageSource(MutableMessageSource activeMessageSource);
+    
+    public  void  setMessageSources(Set<MutableMessageSource>  availableMessageSources);
+
+  
+
+### Proposed API
+
+1.  `setPreferredLocale()`: sets the preferred locale of the user, defaults to English if the user has no preferred locale.
+    
+2.  `getMessages(translation)` : takes argument `es-mx` and returns all the messages for that locale.
+    
+3.  `translate(messageVariableName, englishMessage, target-language)` : takes a message and translates it to the target language, if target language is not found, it returns the englishMessage
+    
+
+  
+  
+
+### Example Code
+
+    import { getCurrentUser } from "@openmrs/esm-api";
```

Comment:
I agree with the above

Reply:
Maybe even the following would be best, since this is going to be so ubiquitous:

```js
import { t } from '@openmrs/esm-l10n'

t("messageId", "First name")
```

Another option: in the frontend web community, there's a convention to use a function `__()` for translations (see [example](https://github.com/webpack/webpack/tree/master/examples/i18n)), which we could also adopt.

Reply:
+1000 to having the default text in code (i.e., using/creating a tool to grep them and create our default English bundle). I'm not a fan of codebases that just put message IDs alone in the code.